### PR TITLE
Admin token validation

### PR DESCRIPTION
**Pull request by shaynafinocchiaro**
# Description
This PR will return a token from the `karavictl admin token` command, adds validation of the admin token at the proxy server, and updates the `make redeploy` target so if the image tags change after deployment, a redeploy will set the new image before a rollout restart. 

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/725|

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Manual testing to validate an appropriate admin token generated from karavictl, new unit tests
- [x] Tested with a custom config signing secret in .karavi/config.json
- [x] Unit Tests
```
# make test
docker run --rm -it -v /root/csm/karavi-authorization/policies:/policies/ openpolicyagent/opa test -v /policies/
/policies/volumes_create_test.rego:
data.karavi.volumes.create.test_small_request_allowed: PASS (2.931454ms)
data.karavi.volumes.create.test_large_request_not_allowed: PASS (483.106µs)
--------------------------------------------------------------------------------
PASS: 2/2
go test -count=1 -cover -race -timeout 30s -short ./...
?       karavi-authorization/cmd/karavictl      [no test files]
?       karavi-authorization/cmd/karavictl/cmd/api/mocks        [no test files]
ok      karavi-authorization/cmd/karavictl/cmd  5.047s  coverage: 67.2% of statements
ok      karavi-authorization/cmd/karavictl/cmd/api      0.153s  coverage: 62.8% of statements
ok      karavi-authorization/cmd/proxy-server   2.998s  coverage: 26.6% of statements
?       karavi-authorization/cmd/role-service   [no test files]
?       karavi-authorization/cmd/sidecar-proxy  [no test files]
?       karavi-authorization/cmd/storage-service        [no test files]
ok      karavi-authorization/cmd/tenant-service 0.111s  coverage: 7.3% of statements
?       karavi-authorization/internal/decision  [no test files]
ok      karavi-authorization/deploy     0.390s  coverage: 84.7% of statements
ok      karavi-authorization/internal/k8s       0.446s  coverage: 51.0% of statements
ok      karavi-authorization/internal/powerflex 10.047s coverage: 91.4% of statements
?       karavi-authorization/internal/storage-service/mocks     [no test files]
ok      karavi-authorization/internal/proxy     13.199s coverage: 70.0% of statements
ok      karavi-authorization/internal/quota     0.493s  coverage: 90.2% of statements
ok      karavi-authorization/internal/role-service      0.097s  coverage: 80.0% of statements
ok      karavi-authorization/internal/role-service/roles        0.138s  coverage: 87.4% of statements
ok      karavi-authorization/internal/role-service/validate     0.431s  coverage: 79.4% of statements
ok      karavi-authorization/internal/sdc       0.099s  coverage: 96.8% of statements
ok      karavi-authorization/internal/storage-service   1.243s  coverage: 92.1% of statements
?       karavi-authorization/internal/tenantsvc/mocks   [no test files]
ok      karavi-authorization/internal/tenantsvc 5.081s  coverage: 79.2% of statements
ok      karavi-authorization/internal/tenantsvc/middleware      0.101s  coverage: 77.7% of statements
ok      karavi-authorization/internal/token     0.116s  coverage: 100.0% of statements
ok      karavi-authorization/internal/token/jwx 0.082s  coverage: 73.1% of statements
?       karavi-authorization/pb [no test files]
ok      karavi-authorization/internal/web       0.088s  coverage: 54.5% of statements
```

### CODE REVIEW

**Review by alikdell — APPROVED** [review]

File: internal/web/middleware.go [hunk 1]
```diff
@@ -124,24 +128,38 @@ func AuthMW(log *logrus.Entry, tm token.Manager) Middleware {
 				parsedToken, err := tm.ParseWithClaims(tkn, JWTSigningSecret, &claims)
 				if err != nil {
 					w.WriteHeader(http.StatusUnauthorized)
-					fwd := forwardedHeader(r)
-					pluginID := normalizePluginID(fwd["by"])
+					fwd := ForwardedHeader(r)
+					pluginID := NormalizePluginID(fwd["by"])
+
+					// an empty plugin ID indicates an admin token
+					if pluginID == "" {
```
Comment (atye) [inline internal/web/middleware.go:135]: In this case when there is an error with the admin token, we should still write an error back to karavictl with `JSONErrorResponse`.

As is, the token could be bad but karavictl wouldn't get an error back.

File: cmd/karavictl/cmd/generate_admin_token.go [hunk 2]
```diff
@@ -53,18 +48,12 @@ func NewAdminTokenCmd() *cobra.Command {
 				reportErrorAndExit(JSONOutput, cmd.ErrOrStderr(), err)
 				return err
 			}
-			if refExpTime <= 0 {
```
Comment (sharmilarama) [inline cmd/karavictl/cmd/generate_admin_token.go:56]: Is this an unreachable condition?

File: cmd/karavictl/cmd/generate_admin_token.go [hunk 3]
```diff
@@ -53,18 +48,12 @@ func NewAdminTokenCmd() *cobra.Command {
...
-				refreshTokenTTL = int64(24 * time.Hour)
-			}
 
 			accExpTime, err := cmd.Flags().GetDuration("access-token-expiration")
 			if err != nil {
 				reportErrorAndExit(JSONOutput, cmd.ErrOrStderr(), err)
 				return err
 			}
-			if accExpTime <= 0 {
```
Comment (sharmilarama) [inline cmd/karavictl/cmd/generate_admin_token.go:65]: Same comment as above.

File: internal/token/jwx/jwx.go [hunk 4]
```diff
@@ -248,3 +253,35 @@ func tokenFromClaims(claims token.Claims) (jwt.Token, error) {
...
+	if req.AccessExpiration <= 0 {
+		req.AccessExpiration = int64(30 * time.Minute)
+	}
+
+	// Generate the token.
+	s, err := token.CreateAdminSecret(tm, token.Config{
+		AdminName:         req.AdminName,
+		Subject:           "admin",
+		Roles:             nil,
+		JWTSigningSecret:  JWTSigningSecret,
```
Comment (sharmilarama) [inline internal/token/jwx/jwx.go:275]: How are you getting the user input for the JWT secret for karavictl admin token command?

File: cmd/karavictl/cmd/generate_admin_token.go (hunk 2 above)
Comment (alikdell) [reply cmd/karavictl/cmd/generate_admin_token.go:56]: Moved the functionality in middleware layer.

File: internal/token/jwx/jwx.go (hunk 4 above)
Comment (alikdell) [reply internal/token/jwx/jwx.go:275]: Either with --password, example: `karavictl admin token --name firstdamin --jwt-signing-secret secret`

OR

omit `--jwt-signing-secret`, cli will prompt for password style input prompt.

File: internal/web/middleware.go (hunk 1 above)
Comment (shaynafinocchiaro) [reply internal/web/middleware.go:138]: fixed

File: internal/token/jwx/jwx.go (hunk 4 above)
Comment (sharmilarama) [reply internal/token/jwx/jwx.go:275]: @alikdell yes I see that in the karavictl command, how is that value being passed here? I see `JWTSigningSecret` is set to default value "secret" in this file.

File: internal/web/middleware.go [hunk 5]
```diff
@@ -127,24 +131,40 @@ func AuthMW(log *logrus.Entry, tm token.Manager) Middleware {
...
+						if err := JSONErrorResponse(w, err); err != nil {
+							log.WithError(err).Println("error parsing the token")
```
Comment (atye) [inline internal/web/middleware.go:140]: This should print "sending json response".

Comment (shaynafinocchiaro) [reply internal/web/middleware.go]: fixed

File: internal/token/jwx/jwx.go (hunk 4 above)
Comment (shaynafinocchiaro) [reply internal/token/jwx/jwx.go]: fixed. Signing secret is now being passed in

File: internal/web/middleware_test.go [hunk 6]
```diff
@@ -13,3 +13,60 @@
...
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+func TestAuthMW(t *testing.T) {
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {})
```
[the remark this answers is not in the document]
Comment (shaynafinocchiaro) [reply internal/web/middleware_test.go]: added unit tests

**Review by alikdell — DISMISSED** [review]

**Review by xuluna — DISMISSED** [review]

**Comment by shaynafinocchiaro** [comment]
run e2e test

**Comment by shaynafinocchiaro** [comment]
run e2e test

**Comment by shaynafinocchiaro** [comment]
run e2e test

**Review by IsaiasA1 — DISMISSED** [review]

**Review by alikdell — APPROVED** [review]

**Comment by shaynafinocchiaro** [comment]
run e2e test

**Review by xuluna — APPROVED** [review]

**Comment by shaynafinocchiaro** [comment]
run e2e test

**Comment by shaynafinocchiaro** [comment]
run e2e test

**Pull request merged**